core/bentley/src/FunctionChain.ts

```diff
@@ -0,0 +1,88 @@
+/*---------------------------------------------------------------------------------------------
+* Copyright (c) Bentley Systems, Incorporated. All rights reserved.
+* See LICENSE.md in the project root for license terms and full copyright notice.
+*--------------------------------------------------------------------------------------------*/
+/** @packageDocumentation
+ * @module Utils
+ */
+
+function prepend(first: () => void, second?: () => void): () => void {
+  if (!second)
+    return first;
+
+  return () => {
+    first();
+    second();
+  };
+}
+
+/** Accumulates a series of functions to be executed in sequence at a later time.
+ * As an example, an object may want to register listeners for various [[BeEvent]]s, and unregister all of those listeners at some point in the future.
+ * The following code illustrates this usage:
+ * ```ts
+ * class Source {
+ *   public readonly onStringEvent = new BeEvent<string>();
+ *   public readonly onNumberEvent = new BeEvent<number>();
+ * }
+ *
+ * class Listener {
+ *   private readonly _unregisterListeners = new FunctionChain();
+ *
+ *   public constructor(source: Source) {
+ *     this._unregisterListeners.append(source.onStringEvent.addListener((str) => this.handleStringEvent(str)));
+ *     this._unregisterListeners.append(source.onNumberEvent.addListener((num) => this.handleNumberEvent(num)));
+ *   }
+ *
+ *   public dispose(): void {
+ *     this._unregisterListeners.callAndClear();
+ *   }
+ *
+ *   private handleStringEvent(str: string) { }
+ *   private handleNumberEvent(num: number) { }
+ * }
+ *```
+ * @beta
+ */
+export class FunctionChain {
```

Comment:
This class is cool, but I'm not sure what virtue it has over BeEvent?

Reply:
Don't need to manually assemble potentially huge list of removal functions. See usage in Viewport.ts and ViewState.ts.

Reply:
I understand what it's used for, but you could use a BeEvent for the same purpose.

Reply:
they're both lists of functions to call. One uses recursion to call and the other uses a vector.

Reply:
Are we talking about FunctionChain, or the example Listener class? Can you elaborate? FunctionChain is used in conjunction with BeEvent, not instead of.

Reply:
Instead of using FunctionChain, use _another_ BeEvent for the "call all these functions to cleanup" object.

Reply:
Sure, but "it's time for me to remove all the event listeners I previously registered" is not really an event. It would be an "event" with exactly one listener - me - that I raise myself to notify myself. Seems pretty awkward to me.

Reply:
well I was going to suggest that instead of using recursion for this class, you should use a vector. But then I realized that that *was* BeEvent. One more class to maintain/explain. But this works too.

Reply:
```ts
export class CleanupList extends BeEvent<() => void> {
  public append(func: () => void): void {
    this.addListener(func);
  }
  public cleanup(): void {
    this.raiseEvent();
    this.clear();
  }
  public get isEmpty(): boolean {
    return 0 === this.numberOfListeners;
  }
}
```

Reply:
I strongly prefer what I have. If you strongly object I can mark it internal or switch to using lists of removal functions.

Reply:
can't think of a single thing to like about the unbounded-recursion approach, but nvm

Reply:
It was not clear to me that you are objecting to the recursion. Would you still object if I switched it to use a list?
I object to using a BeEvent to represent something that's not an event.

Reply:
Yes, I think a class called `FunctionChain` sitting in bentleyjs-core is very likely to get misused by uninformed programmers thinking they're clever. I don't know if there's a "stack" limit on V8 or some other javascript engine, but if there were 1000's of entries who knows. Other than the name "event", I can't see why my approach above isn't better in all respects.

Mainly, I dislike added cruft into that lowest-of-levels package.

Reply:
10,000 minimum recursion depth for modern Javascript engines - but that certainly doesn't mean it's a good idea to get anywhere near that limit.
I'll remove this class and switch to lists of functions.